command/agent/agent_endpoint.go

```diff
@@ -45,9 +56,19 @@ func (s *HTTPServer) AgentReload(resp http.ResponseWriter, req *http.Request) (i
 		return nil, nil
 	}
 
-	errCh := make(chan error, 0)
+	// Fetch the ACL token, if any, and enforce agent policy.
+	var token string
+	s.parseToken(req, &token)
+	acl, err := s.agent.resolveToken(token)
+	if err != nil {
+		return nil, err
+	}
+	if acl != nil && !acl.AgentWrite(s.agent.config.NodeName) {
+		return nil, permissionDeniedErr
+	}
```

Comment:
This logic to check if the token has agent read/write access is repeated a few times in here - do you think it's worth pulling this out to a function similar to `parseToken`?